network/p2p/middleware/middleware.go

```diff
@@ -482,6 +489,17 @@ func (m *Middleware) handleIncomingStream(s libp2pnetwork.Stream) {
 		return
 	}
 
+	streamCount := p2putils.CountStreams(m.libP2PNode.Host(), remotePeer, s.Protocol(), libp2pnetwork.DirInbound)
```

Comment:
Assume `m.unicastMaxStreamsPerPeer = x` and a remote peer concurrently making `y > x` streams to the current peer. Can it result in all `y` streams being deemed over the limit and closed? Ideally, we want `x` out of `y` inbound streams to pass through and `x - y` ones dropped and closed. If that is the case, should we consider a mutual exclusion on handing streams from the same remote peer? i.e., streams from the same remote peer are handled sequentially.

Reply:
I did some testing of this. When `y` is slightly larger than `x`, then usually about 90% of the expected streams are handled correctly. This is about the same as the serial case so the difference is most likely flakiness in the network setup.

When `y` is much larger than `x`, the % false positives grows. 
* at `y == x*4`, ~80% of the expected streams are handled correctly.
* at `y == x*16`, ~60% of the expected streams are handled correctly.

As you noted, this is because there's a race condition between when the stream is opened and when the check occurs in `handleIncomingStream`. Ideally, the add and check would happen atomically so peers would be able to open precisely `m.unicastMaxStreamsPerPeer` streams. 

However, I think the current behavior is OK for a few reasons:
* The number of streams allowed is always less than `m.unicastMaxStreamsPerPeer`, so a malicious peer cannot circumvent the limit.
* When the number of streams is close to the limit, the limit is enforced fairly precisely. This would mean an honest node that's opening slightly larger number of streams, would be treated fairly.
* As the number of concurrently opened streams far exceeds the limit, the percent of rejected streams increases, effectively penalizing bad behavior more.